src/libs/TransactionUtils.ts

```diff
@@ -0,0 +1,358 @@
+import CONST from '../CONST';
+import ONYXKEYS from '../ONYXKEYS';
+import DateUtils from './DateUtils';
+import * as NumberUtils from './NumberUtils';
+import {RecentWaypoints, ReportAction, Transaction} from '../types/onyx';
+import {Receipt, Comment} from '../types/onyx/Transaction';
+import {OriginalMessageIOU} from '../types/onyx/OriginalMessage';
+
+type TransactionChanges = Partial<Transaction> & {comment?: string};
+
+type OriginalMessage = OriginalMessageIOU['originalMessage'];
+
+type Waypoints = Record<string, RecentWaypoints>;
+
+let allTransactions: OnyxCollection<Transaction> = {};
+
+Onyx.connect({
+    key: ONYXKEYS.COLLECTION.TRANSACTION,
+    waitForCollectionCallback: true,
+    callback: (val) => {
+        if (!val) {
+            return;
+        }
+        allTransactions = {...val};
+    },
+});
+
+/**
+ * Optimistically generate a transaction.
+ *
+ * @param amount – in cents
+ * @param currency
+ * @param reportID
+ * @param [comment]
+ * @param [created]
+ * @param [source]
+ * @param [originalTransactionID]
+ * @param [merchant]
+ * @param [receipt]
+ * @param [filename]
+ * @param [existingTransactionID] When creating a distance request, an empty transaction has already been created with a transactionID. In that case, the transaction here needs to have it's transactionID match what was already generated.
+ * @param [category]
+ * @param [tag]
+ * @param [billable]
+ */
+function buildOptimisticTransaction(
+    amount: number,
+    currency: string,
+    reportID: string,
+    comment = '',
+    created = '',
+    source = '',
+    originalTransactionID = '',
+    merchant = '',
+    receipt: Receipt = {},
+    filename = '',
+    existingTransactionID: string | null = null,
+    category = '',
+    tag = '',
+    billable = false,
+): Transaction {
+    // transactionIDs are random, positive, 64-bit numeric strings.
+    // Because JS can only handle 53-bit numbers, transactionIDs are strings in the front-end (just like reportActionID)
+    const transactionID = existingTransactionID ?? NumberUtils.rand64();
+
+    const commentJSON: Comment = {comment};
+    if (source) {
+        commentJSON.source = source;
+    }
+    if (originalTransactionID) {
+        commentJSON.originalTransactionID = originalTransactionID;
+    }
+
+    // For the SmartScan to run successfully, we need to pass the merchant field empty to the API
+    const defaultMerchant = !receipt || Object.keys(receipt).length === 0 ? CONST.TRANSACTION.DEFAULT_MERCHANT : '';
+
+    return {
+        transactionID,
+        amount,
+        currency,
+        reportID,
+        comment: commentJSON,
+        merchant: merchant || defaultMerchant,
+        created: created || DateUtils.getDBTime(),
+        pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+        receipt,
+        filename,
+        category,
+        tag,
+        billable,
+    };
+}
+
+function hasReceipt(transaction: Transaction | null): boolean {
```

Comment:
couldn't we do like that ?
```suggestion
function hasReceipt(transaction?: Transaction): boolean {
```